src/lavinmq/amqp/client.cr

```diff
@@ -410,23 +410,57 @@ module LavinMQ
         @vhost.rm_connection(self)
       end
 
-      private def close_socket
+      private def close_socket(force : Bool = false)
         @running = false
-        @socket.close
-        @log.debug { "Socket closed" }
+        if force && (socket = @socket.as?(Socket))
+          # Socket#close_write will close on "kernel level" which will ensure
+          # we abort any blocking writes. Socket#close tries to flush the socket
+          # etc, meaning that we may block for a long time (or forever for broken
+          # clients)
+          socket.close_write
+          socket.close
+        else
+          @socket.close
+        end
       rescue ex
         @log.debug { "#{ex.inspect} when closing socket" }
       end
 
-      def close(reason = nil)
+      def close(reason = nil, timeout force_close_timeout : Time::Span = 10.seconds)
         reason ||= "Connection closed"
-        @log.info { "Closing, #{reason}" }
-        send AMQP::Frame::Connection::Close.new(320_u16, "CONNECTION_FORCED - #{reason}", 0_u16, 0_u16)
-        @running = false
+        @log.info { "Closing: #{reason}" }
+
+        # We spawn fibers to make #close non-blocking. Required for e.g. http API calls.
+        # It's probably good for VHost#close too which will close all it's connections.
+        # If we run into problems with too many fibers we can probably make one timeout fiber
+        # handling all fibers we're about to close.
```

Comment:
Could we avoid fibers by using `Socket#write_timeout` and `Socket#read_timeout`? Then on `IO::TimeoutError` here and in the read_loop we force close the socket?

Reply:
Yeah, maybe that could be an option. Will give it a try!

Reply:
Seems to work!